aws/rust-runtime/aws-inlineable/src/glacier_interceptors.rs

```diff
@@ -0,0 +1,454 @@
+
+        // Override the signing payload with this precomputed hash
+        let mut signing_config = cfg
+            .get::<SigV4OperationSigningConfig>()
+            .ok_or("SigV4OperationSigningConfig not found")?
+            .clone();
+        signing_config.signing_options.payload_override =
+            Some(SignableBody::Precomputed(content_sha256));
+        cfg.put(signing_config);
+        Ok(())
+    }
+}
+
+/// Adds a glacier tree hash checksum to the HTTP Request
+///
+/// This handles two cases:
+/// 1. A body which is retryable: the body will be streamed through a digest calculator, limiting memory usage.
+/// 2. A body which is not retryable: the body will be converted into `Bytes`, then streamed through a digest calculator.
+///
+/// The actual checksum algorithm will first compute a SHA256 checksum for each 1MB chunk. Then, a tree
+/// will be assembled, recursively pairing neighboring chunks and computing their combined checksum. The 1 leftover
+/// chunk (if it exists) is paired at the end.
+///
+/// See <https://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html> for more information.
+pub(crate) async fn add_checksum_treehash(
+    request: &mut Request<SdkBody>,
+) -> Result<(), byte_stream::error::Error> {
+    let cloneable = request.body().try_clone();
+    let body_to_process = if let Some(cloned_body) = cloneable {
+        // we can stream the body
+        cloned_body
+    } else {
+        let body = std::mem::replace(request.body_mut(), SdkBody::taken());
+        let loaded_body = ByteStream::new(body).collect().await?.into_bytes();
+        *request.body_mut() = SdkBody::from(loaded_body.clone());
+        SdkBody::from(loaded_body)
+    };
+    let (full_body, hashes) = compute_hashes(body_to_process, MEGABYTE).await?;
+    let tree_hash = hex::encode(compute_hash_tree(hashes));
+    let complete_hash = hex::encode(full_body);
+    if !request.headers().contains_key(TREE_HASH_HEADER) {
+        request.headers_mut().insert(
+            HeaderName::from_static(TREE_HASH_HEADER),
+            tree_hash.parse().expect("hash must be valid header"),
+        );
+    }
+    if !request.headers().contains_key(X_AMZ_CONTENT_SHA256) {
+        request.headers_mut().insert(
+            HeaderName::from_static(X_AMZ_CONTENT_SHA256),
+            complete_hash.parse().expect("hash must be valid header"),
+        );
+    }
+    Ok(())
+}
+
+const MEGABYTE: usize = 1024 * 1024;
+async fn compute_hashes(
+    body: SdkBody,
+    chunk_size: usize,
+) -> Result<(Digest, Vec<Digest>), byte_stream::error::Error> {
+    let mut hashes = vec![];
+    let mut remaining_in_chunk = chunk_size;
+    let mut body = ByteStream::new(body);
+    let mut local = Context::new(&SHA256);
+    let mut full_body = Context::new(&SHA256);
+    let mut segmented = SegmentedBuf::new();
+    while let Some(data) = body.try_next().await? {
+        segmented.push(data);
+        while segmented.has_remaining() {
```

Comment:
Just to check, once `segmented` has NO remaining, what will happen when `segmented.push(data)` is called in the next iteration of `while let Some(data) = ...`, assuming the full `segmented` will continue being used in the next iteration?

Reply:
I simplified this by moving the request body reading into the orchestrator.